Fix expconf when empty (unspecified) DataType

reszelaz opened the pull request:
Empty _DataType_ field in expconf is considered as unspecified `data_type` configuration. Pop this configuration key so the _downstream_ software can consider it as unspecified and eventually use default values. Setting it to `None` could be reserved for unknown or erroneous conditions.

Fixes #1070.

@julianofjm could confirm that it fixes your problem? How important is this bug to you? Do you need it in an official release? Or merging into develop would be enough?

julianofjm commented:
Hi @reszelaz . Thanks a lot for the reply. It is not a critical issue for us. We found it out by chance when one of the beamlines configured a measurement group with empty DataType column. Filling in the column solves the issue temporarily, so that is fine to just merge into develop and have this fix in a future release.

I tried to confirm it but I get an error when trying to set the dataType column via expconf:

`Traceback (most recent call last):

  File "/sardana/src/sardana/taurus/qt/qtgui/extra_sardana/measurementgroup.py", line 595, in setData
    item.setData(index, qvalue)
  File "/sardana/src/sardana/taurus/qt/qtgui/extra_sardana/measurementgroup.py", line 363, in setData
    ch_data[key] = data
UnboundLocalError: local variable 'data' referenced before assignment`

Apart from that, using the old measurement group configured before containing a channel with empty dataType seems that the problem still persists, I couldn't see the channel with empty dataType into the HDF file.

One more information is that I could only reproduce this error when creating new measurement groups. Using measurement groups created by sar_demo ("mntgrp01", for example) the issue doesn't appear and the channel is saved, even with dataType column empty.

reszelaz commented:
Thanks for the feedback @julianofjm !

> I get an error when trying to set the dataType column via expconf:

I just fixed this.

I also fixed another bug that was happening when entering to "Data Type" editor where no data type was set and then leaving the editor without setting any data type.

> Apart from that, using the old measurement group configured before containing a channel with empty dataType seems that the problem still persists, I couldn't see the channel with empty dataType into the HDF file.

Yes, it is the case. Before, when a user was leaving empty "Data Type" field in expconf, behind the scene this was setting the `data_type` channel's configuration parameter to `None`. This PR changes this behavior, so now the expconf simply removes this configuration parameter indicating that it is unknown.

This way, if the configuration parameter is not there, the recorder (or ony other part of the system relying on this configuration) may apply its own choice to assign one. I propose to reserve `data_type=None` for indicating abnormal/erroneuous cases.

Based on that, the measurement groups where one explicitly removed the "Data Type" configuration using the expconf before this change are in king of _broken_ state. One would need to set a data type by hand to recover them. But, this is only for the cases where one explicitally removed this configuration parameters. By default, when channel was (and is) added to the expconf it contains the corresponding data type configuration parameter. So I think it is not that bad and we could eventually leave it like this. I suspect that there are not that many users touching this configuration parameter.. What do you think?

> One more information is that I could only reproduce this error when creating new measurement groups. Using measurement groups created by sar_demo ("mntgrp01", for example) the issue doesn't appear and the channel is saved, even with dataType column empty.

The measurement groups created with `sar_demo` are created using the `defmeas` macro. The measurement group created in this way is directly created on the server without applying measurement group configuration. Then it misses the `data_type` configuration paramter. In contrary, when a measurement group is created in expconf, first, it is created on the server and the the expconf applies a full measurement group configuration which contains also the `data_type` parameter. Maybe at some point we should think of having this unified...

julianofjm commented:
Hi @reszelaz 
Thanks! 
I tested your new modification and it works fine. I didn't get the error any more (to set the dataType) and when forcing the dataType column to be empty the channel is still saved into the HDF file, so I think we can consider this issue solved.
I also agree that is fine if old measurements stay in a broken state because of this empty configuration, I saw only once this parameter changed/removed in our measurement group configurations and now we are aware of this.
As I said, since it is not very critical, it is okay for us to wait the next official release to have this fix.
Thanks again and regards,
Juliano

reszelaz commented:
Since @julianofjm confirms that the bug is fixed and this is a minor change I proceed with the merge directly.
